_posts/zh/2021-03-05-ALTF4-26.md

```diff
@@ -0,0 +1,89 @@
+
+{% capture quote_text %}
+**玩家：**嘿，你好！我可以加入异星工厂服务器吗？<br><br>
+
+**服务器：**当然没问题！这是目前的地图，下载吧。[地图在附件]<br><br>
+
+**服务器：**你出现在 x=0, y=3 <br><br>
+
+**服务器：**你的伙伴“potatoman”将他的物流筛选槽 33 设置为“高速传送带”。你把它们也设置为“高速传送带”，并进一步模拟<br><br>
+
+**服务器：**你的伙伴“potatoman”向右移动 3 格<br><br>
+
+**玩家：**我向左移动 4 格<br><br>
+
+**服务器：**已确认，转发中<br><br>
+{% endcapture %}
+
+{% include quote.html text='<img src="https://media.alt-f4.blog/ALTF4/26/factorioserverjoin.jpg" alt="用聊天记录来说明异星工厂服务器如何工作的示意图">' translation=quote_text vertical='true' author='Factorio Server' %}
+
+异星工厂使用了[确定性算法](https://zh.wikipedia.org/wiki/%E7%A1%AE%E5%AE%9A%E6%80%A7%E7%AE%97%E6%B3%95)，当给定相同的输入时，这类算法将产生相同的输出。这意味着结果中不存在随机性，在如异星工厂的情况下正需如此。当异星工厂的多个实例运行时，需要采用完全确定性的算法，这样所有的实例都以锁步算法运行并保持同步。使用完全确定性算法的原因是，如果你有产生随机输出的函数，你就不能使用锁步架构，因为如果处理事件的函数对每个客户端，每次给出的结果都不一样，整个系统就会被搞砸。一个完全确定性的算法是由以下定义的：
+
+* 除算法的输入外，不得使用任何其他数据。不允许使用的数据：随机数、存储的磁盘数据、全局变量、计时器（即程序启动后的时间）
+* 算法的运行不能是时间敏感的
+
+一个反例是，如果一个程序的多个实例写入一个 Excel 电子表格，而另一个程序将读取表格的最后一行。这将使程序具有时间敏感性，因为如果一个写程序的实例延迟了几秒钟，就会产生完全不同的 Excel 行顺序，使读取最后一行的程序获得完全不同的输入。
+
+锁步和完全确定性算法的一个例子是，客户端拍下来的蓝图。当你点击一个蓝图将蓝图导入共享库时，蓝图图标就不会再变灰了，比如下面的右图。这是因为当你点击它们时，你选择将它们转移到游戏的共享库中。然后当你把它放置在某个地方时，你的客户端就会告诉服务器，你把蓝图放置在某个XY坐标上。然后服务器会告诉所有其他连接的客户端，它已经被放置在这些相同的坐标上。然后每个单独的客户端模拟所有的机器人从他们的机器人端口出来，获得资源，放置他们所拥有的实体，然后回到他们选择的机器人端口。所有的客户端都是自己模拟的，没有任何进一步的输入，并且使用前述的完全确定性算法以相同的方式进行模拟。
+
+{% include compare.html id="0" old='https://media.alt-f4.blog/ALTF4/26/blueprints-not-imported.jpg' new='https://media.alt-f4.blog/ALTF4/26/blueprints-imported.jpg' width='800px' %}
+
+失去同步（'Desync'）是指两台计算机根据完全确定性算法应该在同一时间做一些事情并获得相同结果，但没有做到。通常情况下，当客户端和服务器在同一时间做同样的事情时，它们会很高兴，因为它们是同步的（'In sync'）。当两个客户端计算更新的结果不同，通常是由于编程错误，就会出现失去 Desync 的情况。请看下面的图片，这是一个 Desync 的例子。如果一个 Mod 作者或场景作者没有很好地管理他们的数据，也可能会导致 Desync。Desync 会强行让你的客户端从服务器上注销，并生成一个 Desync 报告，这是开发人员用来调查这些 Desync 的东西。
+
+{% capture quote_text %}
+**玩家：**嘿，我计算的 Tick 33859 时刻净功率是 348。对吗？<br><br>
+
+**服务器：**什么？我算的是 936。你搞错了。我把地图发给你，然后断开连接，你可以稍后重连<br><br>
+{% endcapture %}
+
+{% include quote.html text='<img src="https://media.alt-f4.blog/ALTF4/26/desync-chat.jpg" alt="用聊天记录来说明发生 Desync 示意图">' translation=quote_text vertical='true' author='异星工厂服务器' %}
+
+你可能会好奇，当机器人在地图上移动时，怎么不会发生 Desync？当然，如果它们都在工作，并选择一些机器人来执行上述任务，不同的客户端可能会选择不同的机器人来执行这些任务，不是吗？不是的。每个客户端总是会在同一时刻选择同一个机器人，因为选择机器人的算法是 *完全确定的*。两列火车从候车区进站？永远是同一列火车，因为这 *也是完全确定的*。一个喷吐虫决定攻击你的采矿前哨的哪个炮塔？也是完全确定的。这些只是几个例子，但游戏中的 **一切** 都是完全确定性的。如果不是这样，你就会在这里有一个 Desync，在那里又有一个 Desync，多人游戏就根本没法玩了。在多人游戏中，解同步的原因有很多，比如机器人建造、虫群的 AI 模拟，最重要的是，Mod 作者自己造成的。
+
+即使你想使用像 `math.random()` 这样简单的方法在你的 Mod 或场景中得到一个 *随机* 数，也会有一致的结果——所有的客户端都会得到该函数的相同结果。这是因为异星工厂的随机数生成器是 *有种子的*。它从预先赋予的一个数字开始，然后用这个数字在之后产生随机数。如果你让每一个客户端都以同样的方式使用种子，你的 *随机* 数也能保持同步。需要注意的是，它是一个 *伪随机数* 生成器，因此不是真正的随机，因为它是用一个预先确定的数字初始化的，这使得结果可以在任何地方重现。参见[这里](https://en.wikipedia.org/wiki/Random_seed)来了解更多关于随机种子的信息。
```

Comment:
https://en.wikipedia.org/wiki/Random_seed this exist in cantone on wikipedia. did you missed, or is it right that you did not changed the link?

Reply:
Cantonese is not an official language of China but is a dialect that differs greatly from Mandarin (both in pronunciation and in writing). For mainland China, there are probably more people who can speak English than Cantonese speakers. So I chose to keep the Wikipedia page in English instead of Cantonese.